docs/adr/adr-006-fraud-service.md

```diff
@@ -0,0 +1,171 @@
+   Share []byte
+   Proof nmt.Proof
+}
+```
+
+In addition, `das.Daser`:
+
+1. Creates a BEFP:
+
+```go
+const (
+   BadEncoding FraudProofType = "BadEncoding"
+)
+
+type BadEncoding struct {
+   Height uint64
+   Shares []*Share
+   Position uint8
+   isRow bool
+}
+```
+
+2. Full node broadcasts BEFP to all light nodes via separate sub-service via proto message:
+
+```proto3
+
+message MerkleProof {
+  int64          total     = 1;
+  int64          index     = 2;
+  bytes          leaf_hash = 3;
+  repeated bytes aunts     = 4;
+}
+
+message Share {
+   bytes Share = 1;
+   MerkleProof Proof = 2;
+}
+
+message BadEnconding {
+   required uint64 Height = 1;
+   repeated Share Shares = 2;
+   uint8 Position = 3;
+   bool isRow = 4;
+}
+```
+
+`das.Daser` imports a data structure that implements `proof.FraudNotifier` interface that uses libp2p.pubsub under the hood:
+
+```go
+// FraudNotifier is a generic interface that sends a `Proof` to all nodes subscribed on the FraudNotifier's topic.
+type FraudNotifier interface {
+   // Notify takes a fraud `Proof` data structure that implements standard BinaryMarshal interface and broadcasts it to all subscribed peers.
+   Notify(ctx context.Context, p Proof)  
+}
+```
+
+```go
+// FraudProofType is a enum type that represents a particular type of fraud proof.
+type FraudProofType string
+
+// Proof is a generic interface that will be used for all types of fraud proofs in the network.
+type Proof interface {
+   Type() FraudProofType
+   Height() (uint64, error)
+   Validate(*header.ExtendedHeader) error
+
+   encoding.BinaryMarshaller
+   encoding.BinaryUnmarshaller
+}
+```
+
+2a. From the other side, light nodes will, by default, subscribe to the BEFP topic and verify messages received on the topic:
+
+```go
+// Subscriber encompasses the behavior necessary to
+// subscribe/unsubscribe from new FraudProofs events from the
+// network.
+type Subscriber interface {
+   // Subscribe allows to subscribe on pub sub topic by it's type
+   Subscribe(ctx context.Context, proofType FraudProofType) (Subscription, error)
+}
+```
+
+```go
+// Subscription returns a valid proof if one is received on the topic.
+type Subscription interface {
+   Proof() (Proof, error)
+   Cancel() error
+}
+
+// FraudSub implements Subscription.
+type FraudSub struct {
+   pubsub *pubsub.PubSub 
+}
+
+func NewFraudSub(p *pubsub.PubSub)(Subscription, error){}
+func(s *FraudSub) Proof() (Proof, error){}
+func(s *FraudSub) Cancel() error{}
+```
+
+```go
+type FraudService struct {
+   fraudSub Subscription
+   fraudNotifier FraudNotifier
+}
+
+func(s *FraudService) Proof() (Proof, error){}
+func(s *FraudSub) Cancel() error{}
+func(f *FraudService) Notify(ctx context.Context, p Proof){}
+```
+### BEFP verification
+Once a light node receives a `BadEncoding` fraud proof, it should:
+* verify that merkle proofs correspond to particular shares. If the merkle proof does not correspond to a share, then the BEFP is not valid.
```

Comment:
Nit, but Merkle should be capitalized everywhere it's used in text.

Reply:
made them capitalized.